### DISCUSSION TITLE
我在学习中遇到的关于WPF MVVM的6个问题

### DISCUSSION THREAD

**Discussion by Case-Closed-X** [BYJRK/DotNet-Discussions#32] [WPF 开发]
1，在`ViewModel`中捕获异常，如何使用`MessageBox`为用户弹出错误提示呢？

2，使用一些对话框如`OpenFileDialog`时，实例化后调用`ShowDialog()`方法，需要传参`Window owner`建立所有权，如何在MVVM中使用类似的对话框呢？

3，如何为`ViewModel`中的属性赋值`OpenFileDialog`对话框取得的`FileName`呢？

4，在一些不需要使用命令的情况下，如初始化`ObservableCollection`数据的时候，如何控制前台进度条的显示与否呢？

5，在`ViewModel`中如何达到在前台选中`TextBox`中文字的目的呢？

6，在`ViewModel`中怎样控制前台`ListBox`的滚动条滚动到任意一项呢？

**Comment by BYJRK** [comment]
都是不错的问题。我先简单说一下，明天再找时间给你一个具体的答复

1-3 要用到 IoC，或者就是将 View 引入到 ViewModel 中（如果不考虑跨平台开发，其实这是可以接受的
4 可以单独弄一个布尔值进行绑定，然后在初始化的（异步）方法的开始与结束调整这个布尔值，前台 Visibility 绑定它
5-6 属于前台效果，请放在 View 中，而不是想办法在 ViewModel 中实现。具体方法可以考虑用附加属性、行为、自定义控件等

**Reply by BYJRK** [reply]
关于 IoC，你需要有一个容器，可以选择 [DependencyInjection](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection)、[Autofac](https://www.nuget.org/packages/Autofac)、[DryIoc](https://www.nuget.org/packages/DryIoc) 等。然后将 `MessageBox.Show()` 等方法抽象成接口，并在容器中注册，然后自己写一个类给出实现。ViewModel 的构造函数中接受一个上述的接口类型，并将传入的对象存到本地字段。之后如果需要使用弹出框，就调用这个对象的方法

对于最后两个问题，能不能简单说一下你想要实现的功能是什么样的逻辑，比如全选和进度条跳转在什么情况下会被触发，根据什么来决定跳转到哪里等

**Reply by Case-Closed-X** [reply]
好的，感谢您的耐心解答。

`TextBox`的话比如说我想提供给用户一个搜索字符串的功能，用户搜索的字符串在`TextBox`中选中以提示用户字符串的位置；

滚动条的话主要是用在用户通过前台的按钮向列表中插入一个新项，我想让`ListBox`的滚动条滚动到插入的这个新项的位置并且选中该项以提示用户他的操作是有效的。

**Reply by BYJRK** [reply]
关于 `TextBox` 的需求，你或许需要写一个行为，或者自定义控件，从而将整个功能封装起来，而不是暴露给 ViewModel，因为这种需要频繁与前台功能打交道的内容是 MVVM 非常不擅长的。举个例子，比如说你想设计一个这样的 `TextBox`（这是 Win UI 3 中的 `TextBox`）：

![TextBox in Win UI 3](https://user-images.githubusercontent.com/11771667/220007723-61f0ba43-c5e8-44cf-9391-bf4292f3cb46.png)

当内容不为空时，会在框中的右侧出现一个”X“，用来让用户点击并清空内容。那么这样的功能，你应该直接把”X“按钮的显示与否、以及它被点击后的功能全部封装在一个自定义 `TextBox` 中，而不是采用 MVVM，将按钮的 `Visibility`、`Command` 绑定到当前界面的 ViewModel。当然了，你的自定义控件本身可以采用传统的 MVC，也可以采用 MVVM（有单独的 ViewModel），这些都没问题。但总之，不要将界面的所有功能全部塞到一个 ViewModel 中，而是要经常考虑将部分功能拆分成小的模块。

至于第二个需求，你可以考虑给 `ListBox` 添加一个附加属性或者行为（自定义控件当然也是可以的），并在其中去注册 `CollectionChanged` 等事件，并想办法得到新增项目的引用，从而让它能够滚动到相应的位置。

这里有一篇文章，写了一个自定义的 `ListView` 来实现这一效果，你可以参考一下：http://www.dutton.me.uk/2013-08-13/auto-scrolling-a-wpf-listview-when-a-new-item-is-added/

**Reply by Case-Closed-X** [reply]
好的，十分感谢

**Comment by jfaidl** [comment]
学了有一小段时间MVVM和WPF，说说我的看法。使用MVVM就是想前后端分离也就是界面和业务逻辑分离，所以尽量避免在ViewModel中去控制View的行为。按照数据驱动的理念改变ViewModel中绑定到View属性值，让View的进行界面逻辑的处理,View的控件效果和行为就在前端代码中写。

1，在ViewModel中捕获异常，如何使用MessageBox为用户弹出错误提示呢？

MVVM框架中Messager可以做到，Messager可以在ViewModel和View之间通信。在ShellView中注册，接收到消息类后处理显示。


2，使用一些对话框如OpenFileDialog时，实例化后调用ShowDialog()方法，需要传参Window owner建立所有权，如何在MVVM中使用类似的对话框呢？
3，如何为ViewModel中的属性赋值OpenFileDialog对话框取得的FileName呢？

2和3是同一个问题，OpenFileDialog方法应该写在前台代码中，获取到FileName后更新FileName值。这里我用到 `UpdateSourceTrigger=Explicit `和 `textBindingExpression.UpdateSource();`
        


4，在一些不需要使用命令的情况下，如初始化ObservableCollection数据的时候，如何控制前台进度条的显示与否呢？

     这个在构造函数或者重写MVVM框架提供的OnLoaded方法。

**Reply by Case-Closed-X** [reply]
好的，十分感谢
